lib/streamlit/runtime/state/query_params_proxy.py

```diff
@@ -0,0 +1,85 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Iterator, List, MutableMapping, Set, Tuple, Union
+
+from streamlit.runtime.metrics_util import gather_metrics
+from streamlit.runtime.state.query_params import missing_key_error_message
+from streamlit.runtime.state.session_state_proxy import get_session_state
+
+
+class QueryParamsProxy(MutableMapping[str, str]):
+    """A stateless singleton that proxies `st.query_params` interactions
+    to the current script thread's QueryParams instance.
+    """
+
+    def __iter__(self) -> Iterator[str]:
+        with get_session_state().query_params() as qp:
+            return iter(qp)
+
+    def __len__(self) -> int:
+        with get_session_state().query_params() as qp:
+            return len(qp)
+
+    @gather_metrics("query_params.get_item")
+    def __getitem__(self, key: str) -> str:
+        with get_session_state().query_params() as qp:
+            return qp[key]
+
+    def __delitem__(self, key: str) -> None:
+        with get_session_state().query_params() as qp:
+            del qp[key]
+
+    @gather_metrics("query_params.set_item")
+    def __setitem__(
+        self, key: str, value: Union[str, Set[str], Tuple[str], List[str]]
+    ) -> None:
+        with get_session_state().query_params() as qp:
+            qp[key] = value
+
+    @gather_metrics("query_params.get_attr")
+    def __getattr__(self, key: str) -> str:
+        with get_session_state().query_params() as qp:
+            try:
+                return qp[key]
+            except KeyError:
+                raise AttributeError(missing_key_error_message(key))
+
+    def __delattr__(self, key: str) -> None:
+        with get_session_state().query_params() as qp:
+            try:
+                del qp[key]
+            except KeyError:
+                raise AttributeError(missing_key_error_message(key))
+
+    @gather_metrics("query_params.set_attr")
+    def __setattr__(
+        self, key: str, value: Union[str, Set[str], Tuple[str], List[str]]
+    ) -> None:
+        with get_session_state().query_params() as qp:
+            qp[key] = value
+
+    @gather_metrics("query_params.get_all")
+    def get_all(self, key: str) -> List[str]:
+        with get_session_state().query_params() as qp:
+            return qp.get_all(key)
+
+    @gather_metrics("query_params.clear")
+    def clear(self) -> None:
+        with get_session_state().query_params() as qp:
+            qp.clear()
+
+    def to_dict(self) -> Dict[str, str]:
```

Comment:
I think we can also add metrics gathering for this command:

```suggestion
    @gather_metrics("query_params.to_dict")
    def to_dict(self) -> Dict[str, str]:
```

Otherwise, we won't be able to track any read usage of `query_params` for users that just escapes into a dict.